[docs] EKS & AKS auto-discovery docs (#17628)
gravitational/teleport: docs/pages/kubernetes-access/discovery/aws.mdx

---
title: Teleport EKS Auto-Discovery (Preview)
description: Auto-Discovery of EKS clusters in AWS cloud.
---

Similarly to [AKS Auto-Discovery](./azure.mdx), EKS Auto-Discovery can automatically 
discover any EKS cluster and enroll it in Teleport if its tags match the 
configured labels. 

(!docs/pages/kubernetes-access/discovery/includes/step-description.mdx!)

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- An AWS Identity with access to EKS clusters.
- One or more EKS clusters running. 

> 

Author: They can run discovery from their Teleport cluster. It's not required to have dedicated machines

## Known limitations

Due to the authentication method that AWS has chosen for EKS clusters, the 

> Teleport process can't configure the necessary access to Kubernetes clusters on its own.

Author: Teleport does not configure the `configmap/aws-auth` at all. It is a limitation of EKS clusters which prevents you from gaining access to an EKS cluster without previously having access to that same cluster. 
For example, when you create an EKS cluster, your IAM role is automatically associated with the `system:masters` group. If no other user has access to the new cluster and your role is deleted, it is not possible to restore access to the cluster.

That said, Teleport cannot grant access to the cluster by itself (using AWS APIS) and requires someone who already has access to the cluster to edit the configmap and append the Teleport role. (i.e. this process must happen for each cluster to enroll)

It happens because the authentication method on EKS clusters requires that the IAM role 
used for authentication exists in an [`aws-auth` Configmap](https://docs.aws.amazon.com/eks/latest/userguide/add-user-role.html) 
within the target cluster. Therefore, Teleport can't edit the Configmap 
without first having access to the cluster.

If Teleport is not running with the same IAM identity that creates all clusters, 
please ensure that in Step 2 you set the necessary permissions. 

## Step 1/3. Set up AWS IAM credentials

For Teleport to discover EKS clusters, the instance
running the Discovery Service requires an IAM policy that allows Teleport to list and 
describe EKS clusters:

```js
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "eks:DescribeCluster",
                "eks:ListClusters"
            ],
            "Resource": "*"
        }
    ]
}      
```

## Step 2/3. Configure EKS cluster authorization

> When the Kubernetes Service uses an IAM role that is different from the one that 

Author: Unfortunately, it's not possible to know that and it's up to the users to know that information

> creates the clusters, you need to edit the `aws-auth` Configmap and append the 

Author: You have them in the step 2

mapping between the Teleport IAM role and the RBAC permissions on each of the 

> discovered clusters.  You can use EKS cluster creation and management tools such 

Author: This is more a piece of information than anything else. It's not required but if you have a large number of clusters to configure it will save you a lot of time

as `eksctl` and Terraform to automatically configure the access.

> For simplicity, it is possible to map the Teleport IAM role into an existing 

Author: I introduced this because when the number of clusters is large, it is relatively simpler to configure as `system:masters` than to configure `ClusterRole` and `ClusterRoleBinding` resources for each cluster. But I can remove it if it causes confusion

---
title: Teleport EKS Auto-Discovery (Preview)
description: Auto-Discovery of EKS clusters in AWS cloud.
---

Similarly to [AKS Auto-Discovery](./azure.mdx), EKS Auto-Discovery can automatically 
discover any EKS cluster and enroll it in Teleport if its tags match the 
configured labels. 

(!docs/pages/kubernetes-access/discovery/includes/step-description.mdx!)

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- An AWS account with permissions to create and attach IAM policies.
- An AWS account with `system:masters` RBAC access to EKS clusters.
- A host to run the Teleport Discovery and Kubernetes services.
- One or more EKS clusters running. 

## Known limitations

Due to the authorization method that AWS has chosen for EKS clusters, it is not 
possible for the Teleport process to configure the necessary permissions to forward 
requests to EKS clusters.
This limitation exists because the authorization method on EKS clusters requires that the IAM role -
used for authentication - exists in the 
[`aws-auth`](https://docs.aws.amazon.com/eks/latest/userguide/add-user-role.html) `ConfigMap`. 

> In this `ConfigMap``, IAM roles are mapped to Kubernetes RBAC users or groups that are then 

Reviewer: ```suggestion

In this `ConfigMap`, IAM roles are mapped to Kubernetes RBAC users or groups that are then 
```
To break up a long paragraph and fix formatting

used by Kubernetes RBAC to grant access permissions. If the mapping does not exist, it results in 
unauthorized requests. Therefore, Teleport cannot edit the `ConfigMap`
without first having access to the cluster.

If Teleport is not running with the same IAM identity that creates all clusters, 
please ensure you configure the required permissions as described in Step 2. 

## Step 1/3. Set up AWS IAM credentials

For Teleport to discover EKS clusters, the instance
running the Discovery Service requires an IAM policy that allows Teleport to list and 
describe EKS clusters:

```js
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "eks:DescribeCluster",
                "eks:ListClusters"
            ],
            "Resource": "*"
        }
    ]
}      
```

## Step 2/3. Configure EKS cluster authorization

When the Kubernetes Service uses an IAM role that is different from the one that 
creates the clusters, you need to edit the `aws-auth` `Configmap` and append the 
mapping between the Teleport IAM role and the RBAC permissions on each of the 
discovered clusters. 

> For simplicity, it is possible to map the Teleport IAM role into an existing 

Reviewer: ```suggestion
For simplicity, it is possible to map the Teleport IAM role onto an existing 
```

- apiGroups:
  - "authorization.k8s.io"
  resources:
  - selfsubjectaccessreviews
  - selfsubjectrulesreviews
  verbs:
  - create
```

### ClusterRoleBinding

Link the previously created `ClusterRole` into a `teleport` RBAC group.

```yaml
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: teleport
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: teleport
subjects:
- kind: Group
  name: teleport
  apiGroup: rbac.authorization.k8s.io
```

### IAM mapping

Finally, edit the `configmap/aws-auth` in the `kube-system` namespace and append 
the following to `mapRoles`. Please replace `{teleport_aws_iam_role}` with the 
appropriate IAM role that Teleport Kubernetes Service will use. This step will 
link the Teleport IAM role into the Kubernetes RBAC group `teleport`.

```yaml
apiVersion: v1
data:
  mapRoles: |
    - groups:
      - teleport
      rolearn: {teleport_aws_iam_role} # e.g. arn:aws:iam::222222222222:role/teleport-role
      username: teleport
```

  </TabItem>
  <TabItem label="system:masters permissions" >
    <Notice type="warning">
Granting the `system:masters` group to the IAM role associated with the Teleport 

> service means granting administrator-level permissions on the Kubernetes cluster.

Reviewer: ```suggestion
service means granting administrator-level permissions on the Kubernetes cluster. **Do not use this method in production.**
```
To make this a bit more explicit

  </Notice>

> To associate the Teleport IAM role into the `system:masters` RBAC group, 

Reviewer: ```suggestion
To associate the Teleport IAM role with the `system:masters` RBAC group, 
```

> please edit the `configmap/aws-auth` in the `kube-system` namespace and append 

Reviewer: ```suggestion
edit the `configmap/aws-auth` in the `kube-system` namespace and append 
```

the following to `mapRoles`. 

```yaml
apiVersion: v1
data:
  mapRoles: |
  ...
    - groups:
      - system:masters
      rolearn: {teleport_aws_iam_role} # e.g. arn:aws:iam::222222222222:role/teleport-role
      username: teleport
```

> Please replace `{teleport_aws_iam_role}` with the appropriate IAM role that 

Reviewer: ```suggestion
Replace `{teleport_aws_iam_role}` with the appropriate IAM role that the
```
Since technical writing contains a lot of imperative-mood sentences, I think the "please" is usually implied.

---
title: Teleport EKS Auto-Discovery (Preview)

> description: Auto-Discovery of EKS clusters in AWS cloud.

Reviewer 2: ```suggestion
description: How to configure auto-discovery of AWS EKS clusters in Teleport.
```

---

> Similarly to [AKS Auto-Discovery](./azure.mdx), EKS Auto-Discovery can automatically 

Reviewer 2: ```suggestion
EKS Auto-Discovery can automatically 
```

discover any EKS cluster and enroll it in Teleport if its tags match the 
configured labels. 

(!docs/pages/kubernetes-access/discovery/includes/step-description.mdx!)

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- An AWS account with permissions to create and attach IAM policies.
- An AWS account with `system:masters` RBAC access to EKS clusters.
- A host to run the Teleport Discovery and Kubernetes services.
- One or more EKS clusters running. 

> ## Known limitations

Reviewer 2: Think it is worth it to link to that aws-auth API ticket from EKS roadmap here? Maybe in an admonition/note or something. And say that once EKS team implements it, we'll be able to improve this in future. Maybe it will drive more upvotes on that ticket as well :)

Author: Added it to the bottom of Known limitations section

used for authentication - exists in the 
[`aws-auth`](https://docs.aws.amazon.com/eks/latest/userguide/add-user-role.html) `ConfigMap`. 

In this `ConfigMap`, IAM roles are mapped to Kubernetes RBAC users or groups that are then 
used by Kubernetes RBAC to grant access permissions. If the mapping does not exist, it results in 
unauthorized requests. Therefore, Teleport cannot edit the `ConfigMap`
without first having access to the cluster.

If Teleport is not running with the same IAM identity that creates all clusters, 
please ensure you configure the required permissions as described in Step 2. 

## Step 1/3. Set up AWS IAM credentials

For Teleport to discover EKS clusters, the instance
running the Discovery Service requires an IAM policy that allows Teleport to list and 
describe EKS clusters:

```js
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "eks:DescribeCluster",
                "eks:ListClusters"
            ],
            "Resource": "*"
        }
    ]
}      
```

## Step 2/3. Configure EKS cluster authorization

> 

Reviewer 2: ```suggestion
```

When the Kubernetes Service uses an IAM role that is different from the one that 
creates the clusters, you need to edit the `aws-auth` `Configmap` and append the 
mapping between the Teleport IAM role and the RBAC permissions on each of the 
discovered clusters. 

For simplicity, it is possible to map the Teleport IAM role onto an existing 

> Kubernetes RBAC group such as `system:masters`, but Teleport recommends setting 

Reviewer 2: ```suggestion
Kubernetes RBAC group such as `system:masters`, but it is recommended setting 
```

up a separate Kubernetes RBAC group with the minimum required permissions.

<Tabs>

>   <TabItem label="Minimum RBAC permissions (recommended)" >

Reviewer 2: We should add a note somewhere in this section on how we recommend users automate this. Maybe say that if they have Terraform/Cloudformation that provisions EKS clusters and configures aws-auth, they can update that to include Teleport-required permissions? Otherwise users reading this might be wondering why automatic discovery requires manual steps for each cluster.

Connect to your target cluster with your credentials and create the following 
resources using `kubectl`.

> ### ClusterRole

Reviewer 2: Wouldn't Teleport create ClusterRole/Binding if aws-auth configmap allows it access?

Author: In the case of AWS, this does not happen. Although it is technically possible, the Kubernetes RBAC role configured in the configmap probably grants the required Teleport permissions instead of permissions to create `ClusterRoles` and `ClusterRoleBindings` for security reasons.

Should I create a small PR to create them if Teleport has permissions?

Reviewer 2: > the configmap probably grants the required Teleport permissions

Maybe I'm missing something but if the configmap grants the required permissions, why do they need to create these role/binding?

> Should I create a small PR to create them if Teleport has permissions?

I think it would be worthwhile, yeah. The less users have to configure themselves the better. Worst case we'll just get an access denied error and fail gracefully.

Author: > Maybe I'm missing something but if the configmap grants the required permissions, why do they need to create these role/binding?

If it grants the required permissions then you don't need to do anything. The issue arises when the `aws-auth` configmap maps the Teleport IAM role to a Kubernetes RBAC group that allows creating ClusterRole and ClusterRoleBindings but does not allow impersonation.

Reviewer 2: Right, I think the way this section structured is a little confusing, mostly because of the split between 2 options. I wonder if restructuring it like this would make more sense for a reader:

1. Say that Teleport needs X, Y and Z permissions in the cluster and there are following options to pick from:
  - If you don't have any roles with these permissions in the cluster, create this cluster role/binding.
  - If you already have a role in the cluster with these permissions, you can use it.
  - You can also use builtin system:masters which has all permissions but not recommended in production.
2. Map Teleport IAM role to the Kube role from ^^^ using `aws-auth`.

Would this flow be clearer you think? @tigrato

  - get
- apiGroups:
  - "authorization.k8s.io"
  resources:
  - selfsubjectaccessreviews
  - selfsubjectrulesreviews
  verbs:
  - create
```

### ClusterRoleBinding

Link the previously created `ClusterRole` into a `teleport` RBAC group.

```yaml
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: teleport
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: teleport
subjects:
- kind: Group
  name: teleport
  apiGroup: rbac.authorization.k8s.io
```

### IAM mapping

Finally, edit the `configmap/aws-auth` in the `kube-system` namespace and append 
the following to `mapRoles`. Replace `{teleport_aws_iam_role}` with the 
appropriate IAM role that Teleport Kubernetes Service will use. This step will 
link the Teleport IAM role into the Kubernetes RBAC group `teleport`.

```yaml
apiVersion: v1
data:
  mapRoles: |
    - groups:
      - teleport
      rolearn: {teleport_aws_iam_role} # e.g. arn:aws:iam::222222222222:role/teleport-role
      username: teleport
```

  </TabItem>
  <TabItem label="system:masters permissions" >
    <Notice type="warning">
Granting the `system:masters` group to the IAM role associated with the Teleport 

> service means granting administrator-level permissions on the Kubernetes cluster.  **Do not use this method in production.**

Reviewer 2: ```suggestion
service means granting administrator-level permissions on the Kubernetes cluster. To follow least privilege principle we do not recommend using this method in production.
```

  </Notice>

To associate the Teleport IAM role with the `system:masters` RBAC group, 
edit the `configmap/aws-auth` in the `kube-system` namespace and append 
the following to `mapRoles`. 

```yaml
apiVersion: v1
data:
  mapRoles: |
  ...
    - groups:
      - system:masters
      rolearn: {teleport_aws_iam_role} # e.g. arn:aws:iam::222222222222:role/teleport-role
      username: teleport
```

Please replace `{teleport_aws_iam_role}` with the appropriate IAM role that 
Teleport Kubernetes Service is using.

  </TabItem>
</Tabs>

At this point, the Teleport IAM role already has the minimum permissions 
to forward requests to the cluster. 

## Step 3/3. Configure Teleport to discover EKS clusters

> Teleport EKS Auto-Discovery requires a valid auth token for the Discovery and 

Reviewer 2: ```suggestion
Teleport EKS Auto-Discovery requires a valid Teleport auth token for the Discovery and 
```

otherwise might not be clear, is it Teleport token or AWS token or something else.
